storagenode/gracefulexit/worker.go

```diff
@@ -4,48 +4,248 @@
+
+			err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitFailed, msg.ExitFailed.GetExitFailureSignature())
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		case *pb.SatelliteMessage_ExitCompleted:
+			worker.log.Info("graceful exit completed.", zap.String("satellite ID", satelliteID.String()))
+
+			err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitSucceeded, msg.ExitCompleted.GetExitCompleteSignature())
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		default:
+			// TODO handle err
+			worker.log.Error("unknown graceful exit message.", zap.String("satellite ID", satelliteID.String()))
+		}
+
+	}
 
-	err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitSucceeded, []byte{})
 	return errs.Wrap(err)
 }
 
+func (worker *Worker) handleFailure(ctx context.Context, transferError pb.TransferFailed_Error, pieceID pb.PieceID, satelliteID storj.NodeID, send func(*pb.StorageNodeMessage) error) {
+	failure := &pb.StorageNodeMessage{
+		Message: &pb.StorageNodeMessage_Failed{
+			Failed: &pb.TransferFailed{
+				OriginalPieceId: pieceID,
+				Error:           transferError,
+			},
+		},
+	}
+
+	sendErr := send(failure)
+	if sendErr != nil {
+		worker.log.Error("unable to send failure.", zap.String("satellite ID", satelliteID.String()))
+	}
+}
+
 // Close halts the worker.
 func (worker *Worker) Close() error {
 	// TODO not sure this is needed yet.
 	return nil
 }
+
+// TODO This comes from piecestore.Endpoint. It should probably be an exported method so I don't have to duplicate it here.
+func (worker *Worker) getHashAndLimit(ctx context.Context, pieceReader *pieces.Reader, limit *pb.OrderLimit) (pieceHash pb.PieceHash, orderLimit pb.OrderLimit, err error) {
+
+	if pieceReader.StorageFormatVersion() == 0 {
+		// v0 stores this information in SQL
+		info, err := worker.store.GetV0PieceInfoDB().Get(ctx, limit.SatelliteId, limit.PieceId)
+		if err != nil {
+			worker.log.Error("error getting piece from v0 pieceinfo db", zap.Error(err))
+			return pb.PieceHash{}, pb.OrderLimit{}, err
+		}
+		orderLimit = *info.OrderLimit
+		pieceHash = *info.UplinkPieceHash
+	} else {
+		//v1+ stores this information in the file
+		header, err := pieceReader.GetPieceHeader()
+		if err != nil {
+			worker.log.Error("error getting header from piecereader", zap.Error(err))
+			return pb.PieceHash{}, pb.OrderLimit{}, err
+		}
+		orderLimit = header.OrderLimit
+		pieceHash = pb.PieceHash{
+			PieceId:   limit.PieceId,
+			Hash:      header.GetHash(),
+			PieceSize: pieceReader.Size(),
+			Timestamp: header.GetCreationTime(),
+			Signature: header.GetSignature(),
+		}
+	}
+
+	return
+}
+
+// verifyPieceHash verifies whether the piece hash matches the locally computed hash.
+func verifyPieceHash(ctx context.Context, limit *pb.OrderLimit, hash *pb.PieceHash, expectedHash []byte) (err error) {
```

Comment:
is there a reason for using this instead of the piecestore.Endpoint's exported VerifyPieceHash method?

Reply:
piecestore.Endpoint.VerifyPieceHash also checks signatures. As far as I know, we don't have the information to check the signature. I could be wrong.

Reply:
I think we actually do have the information to check the signature. We should be able to get the new node's peer ID from the connection. Then we should be able to get the signee from the peer ID and check the signature with that.

Reply:
added signature verification